dipy/reconst/dki_micro.py

```diff
@@ -0,0 +1,489 @@
+           characterization with diffusional kurtosis imaging.
+           Neuroimage 58(1):177-88. doi: 10.1016/j.neuroimage.2011.06.006
+    """
+    shape = dki_params.shape[:-1]
+
+    # select voxels where to applied the single fiber model
+    if mask is None:
+        mask = np.ones(shape, dtype='bool')
+    else:
+        if mask.shape != shape:
+            raise ValueError("Mask is not the same shape as dki_params.")
+        else:
+            mask = np.array(mask, dtype=bool, copy=False)
+
+    # check or compute awf values
+    if awf is None:
+        awf = axonal_water_fraction(dki_params, sphere, mask=mask)
+    else:
+        if awf.shape != shape:
+            raise ValueError("awf array is not the same shape as dki_params.")
+
+    # Initialize hindered and restricted diffusion tensors
+    EDT = np.zeros(shape + (12,))
+    IDT = np.zeros(shape + (12,))
+
+    # Generate matrix that converts apparant diffusion coefficients to tensors
+    B = np.zeros((sphere.x.size, 6))
+    B[:, 0] = sphere.x * sphere.x  # Bxx
+    B[:, 1] = sphere.x * sphere.y * 2.  # Bxy
+    B[:, 2] = sphere.y * sphere.y   # Byy
+    B[:, 3] = sphere.x * sphere.z * 2.  # Bxz
+    B[:, 4] = sphere.y * sphere.z * 2.  # Byz
+    B[:, 5] = sphere.z * sphere.z  # Bzz
+    pinvB = np.linalg.pinv(B)
+
+    # Compute hindered and restricted diffusion tensors for all voxels
+    evals, evecs, kt = split_dki_param(dki_params)
+    for idx in ndindex(shape):
+        if not mask[idx]:
+            continue
+        # sample apparent diffusion and kurtosis values
+        di = apparent_diffusion_coef(vec_val_vect(evecs[idx], evals[idx]),
+                                     sphere)
+        ki = apparent_kurtosis_coef(dki_params[idx], sphere)
+
+        # Convert apparent diffusion and kurtosis values to apparent diffusion
+        # values of the hindered and restricted diffusion
+        edi = di * (1 + np.sqrt(ki*awf[idx] / (3.0-3.0*awf[idx])))
+        idi = di * (1 - np.sqrt(ki * (1.0-awf[idx]) / (3.0*awf[idx])))
+
+        # generate hindered and restricted diffusion tensors
+        edt = eig_from_lo_tri(np.dot(pinvB, edi))
+        idt = eig_from_lo_tri(np.dot(pinvB, idi))
+        EDT[idx] = edt
+        IDT[idx] = idt
+
+    return EDT, IDT
+
+
+def dkimicro_prediction(params, gtab, S0=1):
+    r""" Signal prediction given the free water DTI model parameters.
+
+    Parameters
+    ----------
+    params : ndarray (x, y, z, 40) or (n, 40)
+    All parameters estimated from the diffusion kurtosis microstructural model.
+        Parameters are ordered as follows:
+            1) Three diffusion tensor's eigenvalues
+            2) Three lines of the eigenvector matrix each containing the
+               first, second and third coordinates of the eigenvector
+            3) Fifteen elements of the kurtosis tensor
+            4) Six elements of the hindered diffusion tensor
+            5) Six elements of the restricted diffusion tensor
```

Comment:
The function `diffusion_components` seems to return 12 parameters for each of these (eigenvalues + eigenvectors). Are those inconsistent with each other? Can we just store 3 eigenvalues for each of these instead of the six tensor elements?